File: src/main/queue/cpp/exec_memory.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <stdexcept>
#include "queue.cpp"

#ifdef __linux__
#include <sys/resource.h>
#elif defined(_WIN32) || defined(_WIN64)
#include <windows.h>
#include <psapi.h>
#endif

using namespace std;

const vector<string> METHOD_NAMES = {
    "addLast",
    "addFirst",
    "add",
    "removeFirst",
    "removeLast",
    "remove",
    "getFirst",
    "getLast",
    "get"
};

long getMemoryUsage() {
    #ifdef __linux__
    struct rusage usage;
    getrusage(RUSAGE_SELF, &usage);
    return usage.ru_maxrss * 1024; // Convertendo de KB para bytes
    #elif defined(_WIN32) || defined(_WIN64)
    PROCESS_MEMORY_COUNTERS_EX pmc;
    GetProcessMemoryInfo(GetCurrentProcess(), (PROCESS_MEMORY_COUNTERS*)&pmc, sizeof(pmc));
    return pmc.PrivateUsage;
    #else
    return 0; // Não suportado em outras plataformas
    #endif
}

void testQueueOperations(const vector<int>& elements, ofstream writers[], int currentLine, int totalLines) {
    int length = elements.size();
    int middle = length / 2;
    const int RUNS = 30;
    vector<vector<long>> allMemoryUsage(METHOD_NAMES.size(), vector<long>(RUNS));

    for (int run = 0; run < RUNS; run++) {
        Queue<int> queue(length + 3);
        
        // Preenche a fila com os elementos iniciais
        for (int num : elements) {
            queue.addLast(num);
        }

        // Medição de memória antes e depois de cada operação
        long beforeMem, afterMem;

        // Testa addLast
        beforeMem = getMemoryUsage();
        queue.addLast(999);
        afterMem = getMemoryUsage();
        allMemoryUsage[0][run] = afterMem - beforeMem;

        // Testa addFirst
        beforeMem = getMemoryUsage();
        queue.addFirst(999);
        afterMem = getMemoryUsage();
        allMemoryUsage[1][run] = afterMem - beforeMem;

        // Testa add (middle)
        beforeMem = getMemoryUsage();
        queue.add(999, middle);
        afterMem = getMemoryUsage();
        allMemoryUsage[2][run] = afterMem - beforeMem;
        
        // Testa removeFirst
        beforeMem = getMemoryUsage();
        queue.removeFirst();
        afterMem = getMemoryUsage();
        allMemoryUsage[3][run] = afterMem - beforeMem;
        
        // Testa removeLast
        beforeMem = getMemoryUsage();
        queue.removeLast();
        afterMem = getMemoryUsage();
        allMemoryUsage[4][run] = afterMem - beforeMem;
        
        // Testa remove (middle)
        beforeMem = getMemoryUsage();
        queue.remove(middle);
        afterMem = getMemoryUsage();
        allMemoryUsage[5][run] = afterMem - beforeMem;
        
        // Testa getFirst (operações get não devem alocar memória)
        beforeMem = getMemoryUsage();
        queue.getFirst();
        afterMem = getMemoryUsage();
        allMemoryUsage[6][run] = afterMem - beforeMem;
        
        // Testa getLast
        beforeMem = getMemoryUsage();
        queue.getLast();
        afterMem = getMemoryUsage();
        allMemoryUsage[7][run] = afterMem - beforeMem;
        
        // Testa get (middle)
        beforeMem = getMemoryUsage();
        queue.get(middle);
        afterMem = getMemoryUsage();
        allMemoryUsage[8][run] = afterMem - beforeMem;
    }

    // Calcula e escreve a mediana para cada método
    for (size_t method = 0; method < METHOD_NAMES.size(); method++) {
        sort(allMemoryUsage[method].begin(), allMemoryUsage[method].end());
        long medianMemory = allMemoryUsage[method][RUNS / 2];
        
        writers[method] << "queue-cpp " << medianMemory << " " << length << endl;
        writers[method].flush();
    }

    // Imprimindo o progresso
    cerr << "\rProcessando linha " << currentLine << "/" << totalLines << flush;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "Uso: " << argv[0] << " <arquivo_entrada>" << endl;
        return 1;
    }

    string inputFile = argv[1];
    string outputDir = "data/results/memory/"; // Alterado para diretório de memória

    // Abre arquivos de saída para cada método
    ofstream writers[METHOD_NAMES.size()];
    for (size_t i = 0; i < METHOD_NAMES.size(); i++) {
        string outputFile = outputDir + METHOD_NAMES[i] + ".data";
        writers[i].open(outputFile, ios::app); // Modo append
        if (!writers[i]) {
            cerr << "\nErro ao abrir arquivo de saída: " << outputFile << endl;
            return 1;
        }
    }

    try {
        ifstream reader(inputFile);
        if (!reader) {
            throw runtime_error("\nNão foi possível abrir o arquivo de entrada: " + inputFile);
        }

        // Captando o total de linhas
        int totalLines = 0;
        string line;
        while (getline(reader, line)) {
            if (!line.empty()) totalLines++;
        }
        reader.clear();
        reader.seekg(0);

        int currentLine = 0;
        while (getline(reader, line)) {
            if (line.empty()) continue;
            currentLine++;
            
            // Converte a linha para vetor de inteiros
            vector<int> elements;
            size_t pos = 0;
            while ((pos = line.find(' ')) != string::npos) {
                string token = line.substr(0, pos);
                elements.push_back(stoi(token));
                line.erase(0, pos + 1);
            }
            if (!line.empty()) {
                elements.push_back(stoi(line));
            }

            testQueueOperations(elements, writers, currentLine, totalLines);
        }

        cerr << "\rConcluído! Processadas " << totalLines << " linhas." << endl;

        // Fecha todos os arquivos
        for (size_t i = 0; i < METHOD_NAMES.size(); i++) {
            writers[i].close();
        }

    } catch (const exception& e) {
        cerr << "\nErro: " << e.what() << endl;
        for (size_t i = 0; i < METHOD_NAMES.size(); i++) {
            if (writers[i].is_open()) {
                writers[i].close();
            }
        }
        return 1;
    }

    return 0;
}

File: src/main/deque/java/ExecMemory.java
import java.io.*;
import java.util.*;

public class ExecMemory {

    private static final String[] METHOD_NAMES = {
        "add_last",
        "add_first",
        "add_middle",
        "remove_first",
        "remove_last",
        "remove_middle",
        "get_first",
        "get_last",
        "get_middle"
    };

    public static void main(String[] args) {

        String inputFile = args[0];
        String outputDir = "data/results/memory/";

        try (BufferedReader reader = new BufferedReader(new FileReader(inputFile))) {
            int totalLines = 0;
            while (reader.readLine() != null) totalLines++;
            reader.close();

            PrintWriter[] writers = new PrintWriter[METHOD_NAMES.length];
            for (int i = 0; i < METHOD_NAMES.length; i++) {
                writers[i] = new PrintWriter(new FileWriter(outputDir + METHOD_NAMES[i] + ".data", true));
            }

            BufferedReader processingReader = new BufferedReader(new FileReader(inputFile));
            String line;
            int currentLine = 0;

            while ((line = processingReader.readLine()) != null) {
                System.out.println("linha "+ currentLine);
                currentLine++;
                if (line.isBlank()) continue;

                System.out.printf("\rProcessando linha %d/%d", currentLine, totalLines);
                System.out.flush();

                int[] elements = Arrays.stream(line.split(" ")).mapToInt(Integer::parseInt).toArray();
                testDequeOperations(writers, elements);
            }

            for (PrintWriter writer : writers) {
                writer.close();
            }

            System.out.printf("\rConcluído! Processadas " + totalLines + " linhas.\n");

            processingReader.close();
        } catch (IOException e) {
            System.err.println("Erro processando arquivo: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static void testDequeOperations(PrintWriter[] writers, int[] elements) {
        int length = elements.length;
        int middle = length / 2;
        int RUNS = 30;
        long[][] allMemory = new long[METHOD_NAMES.length][RUNS];

        Deque deque = new Deque(length + 3);
        for (int num : elements) {
            deque.addLast(num);
        }
        for (int run = 0; run < RUNS; run++) {
            

            // Executa o garbage collector antes dos testes
            System.gc();
            try {
                Thread.sleep(100);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }

            // Testa addLast
            long memoryBefore = getUsedMemory();
            deque.addLast(999);
            long memoryAfter = getUsedMemory();
            allMemory[0][run] = memoryAfter - memoryBefore;

            // Testa addFirst
            memoryBefore = getUsedMemory();
            deque.addFirst(999);
            memoryAfter = getUsedMemory();
            allMemory[1][run] = memoryAfter - memoryBefore;
            deque.removeFirst();

            // Testa add (middle)
            memoryBefore = getUsedMemory();
            deque.add(999, middle);
            memoryAfter = getUsedMemory();
            allMemory[2][run] = memoryAfter - memoryBefore;
            deque.remove(middle);

            // Testa removeFirst
            deque.addFirst(999);
            memoryBefore = getUsedMemory();
            deque.removeFirst();
            memoryAfter = getUsedMemory();
            allMemory[3][run] = memoryBefore - memoryAfter;

            // Testa removeLast
            deque.addLast(999);
            memoryBefore = getUsedMemory();
            deque.removeLast();
            memoryAfter = getUsedMemory();
            allMemory[4][run] = memoryBefore - memoryAfter;

            // Testa remove (middle)
            deque.add(999, middle);
            memoryBefore = getUsedMemory();
            deque.remove(middle);
            memoryAfter = getUsedMemory();
            allMemory[5][run] = memoryBefore - memoryAfter;

            // Testa getFirst
            memoryBefore = getUsedMemory();
            deque.getFirst();
            memoryAfter = getUsedMemory();
            allMemory[6][run] = memoryAfter - memoryBefore;

            // Testa getLast
            memoryBefore = getUsedMemory();
            deque.getLast();
            memoryAfter = getUsedMemory();
            allMemory[7][run] = memoryAfter - memoryBefore;

            // Testa get (middle)
            memoryBefore = getUsedMemory();
            deque.get(middle);
            memoryAfter = getUsedMemory();
            allMemory[8][run] = memoryAfter - memoryBefore;
        }

        for (int method = 0; method < METHOD_NAMES.length; method++) {
            Arrays.sort(allMemory[method]);
            long medianMemory = allMemory[method][RUNS / 2];

            writers[method].println("deque-java " + medianMemory + " " + length);
            writers[method].flush();
        }
    }

    private static long getUsedMemory() {
        Runtime runtime = Runtime.getRuntime();
        return runtime.totalMemory() - runtime.freeMemory();
    }
}
